crates/ty_python_semantic/src/types.rs

```diff
@@ -1172,7 +1125,11 @@ impl<'db> Type<'db> {
             | Type::ClassLiteral(_)
             | Type::SpecialForm(_)
             | Type::IntLiteral(_) => self,
-        }
+        };
+
+        seen_types.remove(&self);
```

Comment:
we do a similar thing in the `inheritance_cycle` query in `class.rs` but there we use an `IndexSet` so that we can just pop the most recently added item off the end of the set:

https://github.com/astral-sh/ruff/blob/90cb0d3a7b0d1b9528854dad64dd17330bfd5f36/crates/ty_python_semantic/src/types/class.rs#L2148

I don't know if that's actually more efficient than using `FxHashSet::remove`, though

Reply:
The follow up PR (WIP) switches this to an `FxOrderSet` anyway, so we can calculate the right de Bruijn index. I'll just change it in this PR for less churn.